### PR TITLE
Confirm the integrity of files with upload resume

### PR DESCRIPTION
**Pull request by djkiourtsis**
When resuming a multipart upload, compare the MD5 of existing parts on aws with corresponding parts from the uploading file to ensure upload integrity.

### CODE REVIEW

**Comment by djkiourtsis** [comment]
[Confirm the integrity of uploaded files](https://trello.com/c/Qicd4GlA/3492-confirm-the-integrity-of-uploaded-files)

File: src/common/AmazonS3Backend.ts [hunk 1]
```diff
@@ -304,11 +304,26 @@ export class AmazonS3Backend implements IBackend {
 
             // aws stores the part md5 in the ETag for multipart uploads
             if (partMD5 === data.Parts[i].ETag) {
-              // TODO add part to completed parts in upload
+
+              // modify current upload object if it exists
+              if (params.Body.s3upload) {
+
+                // ignore part if it has been or is being uploaded
+                if (params.Body.s3upload['completeInfo'][partNumber]) {
```
Comment (abannerjee) [inline src/common/AmazonS3Backend.ts]: Seeing the following warning:
`[312, 42]: object access via string literals is disallowed`

We can fix the warning by changing this line to `params.Body.s3upload.completeInfo[partNumber]`

File: src/common/AmazonS3Backend.ts [hunk 2]
```diff
@@ -304,11 +304,26 @@ export class AmazonS3Backend implements IBackend {
...
+                  return;
+                }
+
+                // update info for complete parts
+                params.Body.s3upload['completeInfo'][partNumber] = {ETag: data.Parts[i].ETag, PartNumber: partNumber};
+                params.Body.s3upload['totalUploadedBytes'] = params.Body.s3upload['totalUploadedBytes'] + data.Parts[i].Size;
+                params.Body.s3upload['doneParts'] = params.Body.s3upload['doneParts'] + 1;
```
Comment (abannerjee) [inline src/common/AmazonS3Backend.ts]: Seeing the following warnings:
```
[317, 38]: object access via string literals is disallowed
[318, 38]: object access via string literals is disallowed
[318, 83]: object access via string literals is disallowed
[319, 38]: object access via string literals is disallowed
[319, 74]: object access via string literals is disallowed
```

Recommend changing these lines to something like:
```
params.Body.s3upload.completeInfo[partNumber] = {
  ETag: data.Parts[i].ETag,
  PartNumber: partNumber
};
params.Body.s3upload.totalUploadedBytes += data.Parts[i].Size;
params.Body.s3upload.doneParts += 1;
```

File: src/common/AmazonS3Backend.ts (hunk 1 above)
Comment (djkiourtsis) [reply src/common/AmazonS3Backend.ts]: This fix works for `totalUploadedBytes` and `doneParts`, but causes an error with `completeInfo`.  Any clue as to why that one is different?
Edit:  The error is no longer appearing for some reason.

**Pull request merged**